src/mozanalysis/frequentist_stats/linear_models.py

```diff
@@ -0,0 +1,412 @@
+      - '0.025': the lower bound of the 95% confidence interval
+      - '0.975': the upper bound of the 95% confidence interval
+      - '0.995': the upper bound of the 99% confidence interval
+    """
+    str_quantiles = ["0.5"]
+    for alpha in alphas:
+        str_quantiles.extend(stringify_alpha(alpha))
+    res = pd.Series(index=sorted(str_quantiles) + ["mean"], dtype="float")
+    dsw = DescrStatsW(branch_data)
+    mean = dsw.mean
+    res["0.5"] = mean  # backwards compatibility
+    res["mean"] = mean
+    for alpha in alphas:
+        low, high = dsw.tconfint_mean(alpha)
+        low_str, high_str = stringify_alpha(alpha)
+        res[low_str] = low
+        res[high_str] = high
+    return res
+
+
+def summarize_univariate(
+    data: pd.Series, branches: pd.Series, branch_list: list[str], alphas: list[float]
+) -> dict[str, pd.Series]:
+    """Univariate inferences (point estimates and confidence intervals) for the
+    mean of each branch's data.
+
+    Parameters:
+    - data (pd.Series): the vector of observations
+    - branches (pd.Series): the vector of branch names. Expected to be the same length
+    as `data`
+    - branch_list (list[str]): the list of branches to perform inferences on
+    - alphas (list[float]): the desired confidence levels
+
+    Returns:
+    - out (dict[str, pd.Series]): a dictionary keyed by branch label with a pandas
+    series as value. The format of the series is described above in
+    `summarize_one_branch`
+    """
+    return {
+        b: summarize_one_branch(data.loc[(branches == b).values], alphas)
+        for b in branch_list
+    }
+
+
+def _make_formula(target: str, ref_branch: str, covariate: str | None = None) -> str:
+    """Makes a formula which defines the model to build. Includes terms for
+    treatment branch and, optionally, a main effect for a (generally pre-experiment)
+    covariate.
+
+    Ex: given `target` of 'Y' and `ref_branch` of control, the formula will be:
+    `Y ~ C(branch, Treatment(reference='control'))`. That is, we want to predict `Y`
+    with treatment branch (`branch`) treated as a categorical predictor with
+    reference level of `'control'`.
+
+    This builds the following linear model (assuming 2 other treatment branches, t1
+    and t2): `Y_i = \beta_0 + \beta_t1*I(branch == 't1') + \beta_t2*I(branch == 't2')`
+    where `I` is the indicator function.
+
+    Inferences on `\beta_t1` are inferences of the average treatment effect (ATE) of
+    branch t1. That is, the confidence interval for `\beta_t1` is the confidence
+    interval for the ATE of branch t1. Similarly for t2 and `\beta_t2`.
+
+    We can incorporate a single (generally pre-experiment) covariate. Adding a covariate
+    of `Ypre` to the above, we'll build the following formula:
+    `Y ~ C(branch, Treatment(reference='control')) + Ypre`
+    which will fit the following linear model:
+
+    `Y_i = \beta_0 + \beta_t1*I(branch == 't1') + \beta_t2*I(branch == 't2') + Ypre_i`
```

Comment:
Why doesn't the covariate get a parameter?  Why can you only have one?

Reply:
1) because I forgot. Edited the formula to add it in. 2) you don't have to have only one, I'm thinking we start with 1 (the pre-treatment version of the metric being studied) for simplicity, and can expand from there. I'd like to see us get to many, but that's much more complex.